File: src/commons/atoms/Node.cc
#include "Node.h"

#include "Hasher.h"
#include "MettaMapping.h"

using namespace atoms;

// -------------------------------------------------------------------------------------------------
// Constructors, destructors , basic operators and initializers

Node::Node(const string& type, const string& name, bool is_toplevel, const Properties& custom_attributes)
    : Atom(type, is_toplevel, custom_attributes), name(name) {
    this->validate();
}

Node::Node(const string& type, const string& name, const Properties& custom_attributes)
    : Atom(type, false, custom_attributes), name(name) {
    this->validate();
}

Node::Node(const Node& other) : Atom(other) { this->name = other.name; }

Node& Node::operator=(const Node& other) {
    Atom::operator=(other);
    this->name = other.name;
    return *this;
}

bool Node::operator==(const Node& other) {
    return (this->name == other.name) && Atom::operator==(other);
}

bool Node::operator!=(const Node& other) { return !(*this == other); }

void Node::validate() const {
    if (this->type == Atom::UNDEFINED_TYPE) {
        Utils::error("Node type can't be '" + Atom::UNDEFINED_TYPE + "'");
    }
    if (this->name.empty()) {
        Utils::error("Node name must not be empty");
    }
}

// -------------------------------------------------------------------------------------------------
// Public API

string Node::to_string() const {
    return "Node(type: '" + this->type + "', name: '" + this->name +
           "', custom_attributes: " + this->custom_attributes.to_string() + ")";
}

string Node::handle() const { return Hasher::node_handle(this->type, this->name); }

string Node::metta_representation(HandleDecoder& decoder) const {
    if (this->type != MettaMapping::SYMBOL_NODE_TYPE) {
        Utils::error("Can't compute metta expression of node whose type (" + this->type + ") is not " +
                     MettaMapping::SYMBOL_NODE_TYPE);
    }
    return this->name;
}

bool Node::match(const string& handle, Assignment& assignment, HandleDecoder& decoder) {
    return this->handle() == handle;
}

void Node::tokenize(vector<string>& output) {
    output.insert(output.begin(), this->name);
    Atom::tokenize(output);
}

void Node::untokenize(vector<string>& tokens) {
    Atom::untokenize(tokens);
    this->name = tokens[0];
    tokens.erase(tokens.begin(), tokens.begin() + 1);
}

File: 3rd_party_slots/python_client/hyperon_das/commons/atoms/node.py
import functools

from hyperon_das.commons.atoms import Atom
from hyperon_das.commons.atoms.handle_decoder import HandleDecoder
from hyperon_das.commons.helpers import error
from hyperon_das.commons.properties import Properties
from hyperon_das.hasher import Hasher
from hyperon_das.query_answer import Assignment


class Node(Atom):
    def __init__(
        self,
        type: str | None = None,
        name: str | None = None,
        is_toplevel: bool = False,
        custom_attributes: Properties | None = None,
        tokens: list[str] | None = None,
        other=None,
    ) -> None:
        if tokens:
            self.custom_attributes = custom_attributes if custom_attributes is not None else Properties()
            self.untokenize(tokens)
        elif other:
            self._name = other.name
            super().__init__(other=other)
        else:
            if not type or not name:
                raise ValueError("'type' and 'name' are required")
            self._name = name
            super().__init__(type, is_toplevel, custom_attributes)

    @property
    def name(self) -> str:
        return self._name

    @name.setter
    def name(self, value: str) -> None:
        self._name = value

        # If the handle has already been calculated, delete it.
        if "handle" in self.__dict__:
            del self.__dict__["handle"]

    def __eq__(self, other: 'Node') -> bool:
        if not isinstance(other, Node):
            return False
        return (
            self.name == other.name
            and self.type == other.type
            and self.is_toplevel == other.is_toplevel
            and self.custom_attributes == other.custom_attributes
        )

    def __ne__(self, other: 'Node') -> bool:
        return not self.__eq__(other)

    def copy_from(self, other: 'Node') -> 'Node':
        super().copy_from(other)
        self.name = other.name
        return self

    def validate(self) -> None:
        if self.type == Atom.UNDEFINED_TYPE:
            error(f"Node type can't be '{Atom.UNDEFINED_TYPE}'")

        if len(self.name) == 0:
            error("Node name must not be empty")

    def to_string(self) -> str:
        return (
            f"Node(type: '{self.type}', name: '{self.name}', custom_attributes: '{self.custom_attributes.to_string()}'"
        )

    def tokenize(self, output: list[str]) -> None:
        super().tokenize(output)
        output.append(self.name)

    def untokenize(self, tokens: list[str]) -> None:
        super().untokenize(tokens)
        self.name = tokens[0]
        del tokens[:1]

    @functools.cached_property
    def handle(self) -> str:
        return Hasher.node_handle(self.type, self.name)

    def metta_representation(self, decoder: 'HandleDecoder') -> str:
        if self.type != "Symbol":
            error(f"Can't compute metta expression of node whose type ({self.type}) is not Symbol")
        return self.name

    def match(self, handle, assignment: 'Assignment', decode: 'HandleDecoder') -> bool:
        return self.handle == handle
